source/features/useful-not-found-page.tsx

```diff
@@ -36,46 +38,156 @@ function parseCurrentURL(): string[] {
 	return parts;
 }
 
-// If the resource was deleted, link to the commit history
-async function addCommitHistoryLink(bar: Element): Promise<void> {
+async function findAndDisplayMissingPart(bar: Element): Promise<void> {
+	// The branch might have been deleted
+	const branchDeleted = await addBranchStatus(bar);
+	// Or the file exists, but on a different branch
+	const branchDoesNotHaveFile = await addDefaultBranchLink(bar);
```

Comment:
When both of these conditions are true (see the [example screenshot](https://user-images.githubusercontent.com/34235681/128588754-d000609d-389f-45e8-b671-8ad957e6a7e4.png) for non existent branch but the file exists on default branch), the `addDefaultBranchLink` function adds the link to the file on the default branch, above the text that says the branch does not exist (added by the `addBranchStatus` function). UI-wise, the order of the text looks wrong to me.

This could be happening because of two reasons:

- We pass the same `bar` instance to both functions, and `bar.after()` keeps adding the elements to the top.
- The `addDefaultBranchLink` function finishes before the `addBranchStatus` function, thus adding its text to the `bar` first.

Possible solutions:

- Combining the two functions and adding the text as one `p` element.
- Returning the bar instance from the `addBranchStatus` function and passing that to the `addDefaultBranchLink` function (might not work if this error happens due to the first reason)
- Not showing the branch does not exist error if the file does exist on another branch.

It would be great if someone could advise me on how to fix this. Thanks!

Reply:
I think we have 2 parallel requests to make:

- does the file exist on default branch, currently
- does the file exist in history

If both are false

- does the file exist in default’s history

If that's also false, the file never existed.

> - does the file exist on default branch, currently

We show "The file only exists on the default branch" or "The branch has been deleted, see the file on the default branch"

> - does the file exist in history

We show "The file was deleted" or "The file was moved"

> - does the file exist in default’s history

Same as above, but change the links to point to default (we don't need to say anything about the missing branch)

Reply:
For this reason, the file should be reorganized in a way to have a single "What happened to this file" function (other than the breadcrumbs, which should not be affected). And then use it as something like:

```js
// Random names, to be improved, pseudo-code
const outcome = await whatHappened(location.href);
if (outcome.type === 'renamed') {
	page.append(<a href={outcome.oldFile}>File</a>, 'was renamed etc etc')
} else if (/* deleted */) {
	page.append(<a href={outcome.oldFile}>File</a>, 'was deleted etc etc')
}